src/execution/operator/order/physical_order.cpp

```diff
@@ -1130,35 +1352,153 @@ class PhysicalOrderMergeTask : public Task {
 			}
 			Merge(*result->payload_data, *left.payload_data, *right.payload_data, next, left_smaller, next_entry_sizes);
 		}
-		D_ASSERT(result->Count() == left.Count() + right.Count());
+		D_ASSERT(result->Count() == l_count + r_count);
 
 		lock_guard<mutex> glock(state.lock);
-		state.sorted_blocks_temp.push_back(move(result));
 		parent.finished_tasks++;
-		if (parent.total_tasks == parent.finished_tasks) {
+		if (parent.finished_tasks == parent.total_tasks) {
 			for (auto &sb : state.sorted_blocks) {
 				sb->UnregisterSortingBlocks();
 				sb->UnregisterPayloadBlocks();
 			}
 			state.sorted_blocks.clear();
-			for (auto &block : state.sorted_blocks_temp) {
-				state.sorted_blocks.push_back(move(block));
+			if (state.odd_one_out) {
+				state.sorted_blocks.push_back(move(state.odd_one_out));
+				state.odd_one_out = nullptr;
+			}
+			for (auto &sorted_block_vector : state.sorted_blocks_temp) {
+				state.sorted_blocks.push_back(
+				    make_unique<SortedBlock>(buffer_manager, *state.sorting_state, *state.payload_state));
+				state.sorted_blocks.back()->AppendSortedBlocks(sorted_block_vector);
 			}
 			state.sorted_blocks_temp.clear();
 			PhysicalOrder::ScheduleMergeTasks(parent, context, state);
 		}
 	}
 
+	//! Compare values within SortedBlocks using a global index
+	int CompareUsingGlobalIndex(SortedBlock &l, SortedBlock &r, const idx_t l_idx, const idx_t r_idx) {
+		D_ASSERT(l_idx < l.Count());
+		D_ASSERT(r_idx < r.Count());
+
+		idx_t l_entry_idx;
+		idx_t l_block_idx;
+		l.GlobalToLocalIndex(l_idx, l_block_idx, l_entry_idx);
+
+		idx_t r_entry_idx;
+		idx_t r_block_idx;
+		r.GlobalToLocalIndex(r_idx, r_block_idx, r_entry_idx);
+
+		auto l_block_handle = buffer_manager.Pin(l.sorting_blocks[l_block_idx].block);
+		auto r_block_handle = buffer_manager.Pin(r.sorting_blocks[r_block_idx].block);
+		data_ptr_t l_ptr = l_block_handle->Ptr() + l_entry_idx * sorting_state.entry_size;
+		data_ptr_t r_ptr = r_block_handle->Ptr() + r_entry_idx * sorting_state.entry_size;
+
+		int comp_res;
+		if (sorting_state.all_constant) {
+			comp_res = memcmp(l_ptr, r_ptr, sorting_state.comp_size);
+		} else {
+			l.SetVarIndices(l_idx);
+			r.SetVarIndices(r_idx);
+			l.PinVarBlocks();
+			r.PinVarBlocks();
+			comp_res = CompareColumns(l, r, l_ptr, r_ptr, sorting_state);
+		}
+		return comp_res;
+	}
+
+	void ComputeIntersection(SortedBlock &l, SortedBlock &r, const idx_t sum, idx_t &l_idx, idx_t &r_idx) {
+		const idx_t l_count = l.Count();
+		const idx_t r_count = r.Count();
+		// cover some edge cases
+		if (sum >= l_count + r_count) {
+			l_idx = l_count;
+			r_idx = r_count;
+			return;
```

Comment:
Are all of these edge cases encountered in the tests?

Reply:
I created `ComputeIntersection` in a separate C++ script to make development quicker. There, I've covered all of these edge cases. Of the first 4 edge cases listed there, I am 100% sure that the last two cases are encountered in the tests (`test_order_parallel.test` covers these).

The first two cases however, are probably not triggered, since the first case is handled in `PhysicalOrderMergeTask`, [line 1294](https://github.com/duckdb/duckdb/blob/abecd554d11110698a71d97863d7dc84859488d0/src/execution/operator/order/physical_order.cpp#L1294), and the second case is avoided by the way the tasks are scheduled in `PhysicalOrder::ScheduleMergeTasks`.

However, if that code changes, these first two edge cases may be triggered again. Would you rather have this code be removed for now?


As for the bunch of if/else statements in the binary search part of `ComputeIntersection`, I am also sure that those are covered. If you remove one of these you can be sure of getting an off-by-one somewhere in `make allunit` ;)

Reply:
That sounds good enough to me. I saw these edge cases and imagined it would be difficult to trigger them in the test suite, but if they have been tested I am fine with the code staying there. I imagine they could get hit in an edge case so these protections are fine. Would simply like to avoid code being checked in that has never been run before :)